mixnet/src/client.rs

```diff
@@ -1,25 +1,111 @@
+use std::collections::VecDeque;
+
+use rand::rngs::OsRng;
 use serde::{Deserialize, Serialize};
 use tokio::sync::mpsc;
 
-use crate::{error::MixnetError, packet::Packet};
+use crate::{error::MixnetError, packet::Packet, poisson::Poisson, topology::MixnetTopology};
 
 /// Mix client implementation that returns packets in Poisson intervals
-pub struct MixClient {}
+pub struct MixClient {
+    packet_rx: mpsc::UnboundedReceiver<Packet>,
+}
+
+struct MixClientRunner {
+    config: MixClientConfig,
+    poisson: Poisson,
+    message_queue: mpsc::Receiver<Box<[u8]>>,
+    real_packet_queue: VecDeque<Packet>,
+    packet_tx: mpsc::UnboundedSender<Packet>,
+}
 
 /// Mix client configuration
 #[derive(Serialize, Deserialize, Clone, Debug)]
-pub struct MixClientConfig {}
+pub struct MixClientConfig {
+    /// Mixnet topology
+    pub topology: MixnetTopology,
+    /// Poisson rate for packet emissions (per minute)
+    pub emission_rate_per_min: f64,
+    /// Packet redundancy for passive retransmission
+    pub redundancy: u8,
+}
+
+const MESSAGE_QUEUE_SIZE: usize = 256;
 
 /// Queue for sending messages to [`MixClient`]
 pub type MessageQueue = mpsc::Sender<Box<[u8]>>;
 
 impl MixClient {
-    pub fn new(_: MixClientConfig) -> Result<(Self, MessageQueue), MixnetError> {
-        todo!()
+    /// Creates a [`MixClient`] and a [`MessageQueue`].
+    ///
+    /// This returns [`MixnetError`] if the given `config` is invalid.
+    pub fn new(config: MixClientConfig) -> Result<(Self, MessageQueue), MixnetError> {
+        let poisson = Poisson::new(config.emission_rate_per_min)?;
+        let (tx, rx) = mpsc::channel(MESSAGE_QUEUE_SIZE);
+        let (packet_tx, packet_rx) = mpsc::unbounded_channel();
+
+        MixClientRunner {
+            config,
+            poisson,
+            message_queue: rx,
+            real_packet_queue: VecDeque::new(),
+            packet_tx,
+        }
+        .run();
+
+        Ok((Self { packet_rx }, tx))
     }
 
     /// Returns a next [`Packet`] to be emitted, if it exists and the Poisson timer is done.
     pub async fn next(&mut self) -> Option<Packet> {
-        todo!()
+        self.packet_rx.recv().await
+    }
+}
+
+impl MixClientRunner {
+    fn run(mut self) {
+        tokio::spawn(async move {
+            let mut delay = tokio::time::sleep(self.poisson.interval(&mut OsRng));
+            loop {
+                let next_deadline = delay.deadline() + self.poisson.interval(&mut OsRng);
+                delay.await;
+                delay = tokio::time::sleep_until(next_deadline);
```

Comment:
Nit: 

```rust
let mut delay = tokio::time::sleep(self.poisson.interval(&mut OsRng));
tokio::pin!(delay);

loop {
   let next_deadline = delay.deadline() + self.poisson.interval(&mut OsRng);
   delay.await;
   delay.as_mut().reset(next_deadline);
}
```

Reply:
Curious, why the pin here?

Reply:
`tokio`'s `Sleep` API, `reset` only exists on `Pin<&mut Self>`, so have to pin here.

Reply:
If we don't use `reset` (as I did), don't we need to pin it?

Reply:
Done in https://github.com/logos-co/nomos-node/pull/616